MaxText/layers/linears.py

```diff
@@ -410,11 +410,28 @@ def wrapper(x, logits, w0, w1, wo):
     return wrapper(inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel)
 
   def dense_matmul(self, inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel):
+    # shape of top_k_weights & top_k_indices: (batch, sequence, exp_per_token)
     top_k_weights, top_k_indices = jax.lax.top_k(gate_logits, self.num_experts_per_tok)
-    flattened_top_k_weights = top_k_weights.reshape(-1, self.num_experts_per_tok)
-
-    softmax_probs = jax.nn.softmax(flattened_top_k_weights.astype(jnp.float32), axis=-1).astype(self.weight_dtype)
-    softmax_probs = softmax_probs.reshape(gate_logits.shape[:-1] + (self.num_experts_per_tok,))
+    softmax_probs = jax.nn.softmax(top_k_weights.astype(jnp.float32), axis=-1).astype(self.weight_dtype)
+
+    # token dropping
+    if self.config.capacity_factor > 0:
+      # calculate expert_capacity = (tokens_per_batch / num_experts) * capacity_factor
+      batch_size, seq_len, _ = inputs.shape
+      tokens_per_batch = batch_size * seq_len * self.num_experts_per_tok
+      expert_capacity = int((tokens_per_batch / self.num_experts) * self.config.capacity_factor)
+      max_logging.log(f"Token dropping if needed, and expert_capacity is {expert_capacity}")
+
+      # calculate expert mask and drop tokens if needed
+      # shape of expert_mask: (batch, sequence, exp_per_token, num_experts)
+      expert_mask = jax.nn.one_hot(top_k_indices, num_classes=self.num_experts)
+      cumsum_expert_mask = jnp.cumsum(jnp.cumsum(expert_mask, axis=1), axis=2)
```

Comment:
Comments here would be helpful although it is tricky how to best explain this concisely (The code is more concise than words). I prefer the name "expert_cumusm_token_count" instead of cumsum_expert_mask for line 428, and maybe expert_token_count for line 429 (open to other ideas though). Perhaps a small example - "If expert 0 is chosen by tokens [1,3] the expert_token_count will have entries [0,1,0,2], which associates each chosen token with an increasing count"

Reply:
SG! Will update the variable name and make a comment there.

Reply:
Updated the variables, and have a small example there. Please let me know if it's unclear.